src/map/core/raw.rs

```diff
@@ -143,3 +157,80 @@ impl<'a, K, V> RawTableEntry<'a, K, V> {
         (self.map, index)
     }
 }
+
+pub(crate) struct ExtractCore<'a, K, V> {
+    map: &'a mut IndexMapCore<K, V>,
+    current: usize,
+    new_len: usize,
+}
+
+impl<K, V> Drop for ExtractCore<'_, K, V> {
+    fn drop(&mut self) {
+        let old_len = self.map.indices.len();
+        let mut new_len = self.new_len;
+        debug_assert!(new_len <= self.current);
+        debug_assert!(self.current <= old_len);
+        debug_assert!(old_len <= self.map.entries.capacity());
+
+        // SAFETY: We assume `new_len` and `current` were correctly maintained by the iterator.
+        // So `entries[new_len..current]` were extracted, but the rest before and after are valid.
+        unsafe {
+            if new_len == self.current {
+                // Nothing was extracted, so any remaining items can be left in place.
+                new_len = old_len;
+            } else if self.current < old_len {
+                // Need to shift the remaining items down.
+                let tail_len = old_len - self.current;
+                let base = self.map.entries.as_mut_ptr();
+                let src = base.add(self.current);
+                let dest = base.add(new_len);
+                src.copy_to(dest, tail_len);
+                new_len += tail_len;
+            }
+            self.map.entries.set_len(new_len);
+        }
+
+        if new_len != old_len {
+            // We don't keep track of *which* items were extracted, so reindex everything.
+            self.map.rebuild_hash_table();
```

Comment:
While it would be nice to reuse `vec::ExtractIf`, I haven't found a way to do this fixup for the indices afterward, because that `vec::ExtractIf` would be holding `&mut` on the entries, excluding even a raw pointer elsewhere.